src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -178,14 +181,84 @@ class SetupPyChrootRequest:
     py2: bool  # Whether to use py2 or py3 package semantics.
 
 
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class SetupKwargs:
+    """The keyword arguments to the `setup()` function in the generated `setup.py`."""
+
+    json_str: str
+
+    def __init__(
+        self, kwargs: Mapping[str, Any], *, address: Address, _allow_banned_keys: bool = False
+    ) -> None:
+        super().__init__()
+        if "version" not in kwargs:
+            raise ValueError(f"Missing a `version` kwarg in the `provides` field for {address}.")
+
+        if not _allow_banned_keys:
+            for arg in {"data_files", "package_dir", "package_data", "packages"}:
+                if arg in kwargs:
+                    raise ValueError(
+                        f"{arg} cannot be set in the `provides` field for {address}, but it was "
+                        f"set to {kwargs[arg]}. Pants will dynamically set the value for you."
+                    )
+
+        # We convert to a `str` so that this type can be hashable. We don't use `FrozenDict`
+        # because it would require that all values are immutable, and we may have lists and
+        # dictionaries as values. It's too difficult/clunky to convert those all, then to convert
+        # them back out of `FrozenDict`.
+        self.json_str = json.dumps(kwargs, sort_keys=True)
+
+    @memoized_property
+    def kwargs(self) -> Dict[str, Any]:
+        return cast(Dict[str, Any], json.loads(self.json_str))
+
+    @property
+    def name(self) -> str:
+        return cast(str, self.kwargs["name"])
+
+    @property
+    def version(self) -> str:
+        return cast(str, self.kwargs["version"])
+
+
+@union
+@dataclass(frozen=True)  # type: ignore[misc]
+class SetupKwargsPluginRequest(ABC):
```

Comment:
The word `Plugin` might be confusing here. This isn't directly related to a pants plugin per se. I think `SetupKwargsRequest` is fine (e.g., you named our implementation `PantsSetupKwargsRequest`, not `PantsSetupKwargsPluginRequest`).

Reply:
The reason I used `Plugin` is to clarify that this is only if users added their own implementation. I'm afraid that `SetupKwargsRequest` would make you think to do `await Get(SetupKwargs, SetupKwargsRequest)`, when normally you'd want to do `await Get(SetupKwargs, ExportedTarget)`.

What do you think?

Reply:
Who is "you" in that sentence? The plugin author wouldn't be doing either of those `Get` s, right?  That's in `run_setup_py` ?

Reply:
"you" is `run_setup_py`. A plugin author is unlikely to do anything more than write a rule `MyCustomKwargsRequest -> SetupKwargs`.

The only places that will call `await Get(SetupKwargs, SetupKwargsPluginRequest)` and `await Get(SetupKwargs, ExportedTarget)` are `run_setup_py.py` and any custom plugins that are doing something unusual like re-implenenting `setup-py` their one way through a custom goal.

Reply:
Right, so I'm less concerned about the naming causing the internal confusion you describe.

I'm still not keen on the `Plugin` designation.